server/src/main/java/org/opensearch/index/store/remote/utils/ConcurrentInvocationLinearizer.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.store.remote.utils;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ExecutorService;
+import java.util.function.Function;
+
+/**
+ * A utility class which can be used to serialize concurrent invocations and to achieve "invoke simultaneously once at any time" semantic
+ *
+ * @param <METHOD_PARAM_TYPE> the method parameter type where this method invocation will be linearized
+ * @param <RET_TYPE>          return type of the method
+ * @opensearch.internal
+ */
+public class ConcurrentInvocationLinearizer<METHOD_PARAM_TYPE, RET_TYPE> {
+    private final ConcurrentMap<METHOD_PARAM_TYPE, CompletableFuture<RET_TYPE>> invokeOnceCache;
+    private final ExecutorService executorService;
+
+    /**
+     * Constructs the object
+     *
+     * @param executorService which will be used to execute the concurrent invocations
+     */
+    public ConcurrentInvocationLinearizer(ExecutorService executorService) {
+        this.invokeOnceCache = new ConcurrentHashMap<>();
+        this.executorService = executorService;
+    }
+
+    /**
+     * @param input    the argument to the method
+     * @param function delegate to actual function/method
+     * @return return value of the function
+     */
+    public CompletableFuture<RET_TYPE> linearize(METHOD_PARAM_TYPE input, Function<METHOD_PARAM_TYPE, RET_TYPE> function) {
+        return invokeOnceCache.computeIfAbsent(input, in -> CompletableFuture.supplyAsync(() -> function.apply(in), executorService))
+            .whenComplete((ret, throwable) -> {
+                // whenComplete will always be executed (when run normally or when exception happen)
+                invokeOnceCache.remove(input);
```

Comment:
Per the contract of `computeIfAbsent`: "The mapping function should not modify this map during computation." This is a little different because the modification happens asynchronously, but it makes me think there could be nasty race conditions here. Also I think the mapping function should have no side effects as I believe it can be invoked even if the item does not get added to the map, which here might result in multiple remove calls happening for a single entry in the map. Maybe that's fine? But it makes me worry about race conditions.

Reply:
yeah my biggest concern here is that it is confusing. but I couldn't think about any race condition here as `whenComplete` is executed synchronously after first stage (mapping function) is complete and it's not within the scope of `computeIfAbsent` mapping function so it's violating any contract. this `whenComplete` just act as a hook to run before returning res to caller. I think any concurrent issues is handler by the concurrent hash map and I cannot think about race conditions causing inconsistent state here so can you think about any ?

Reply:
Yeah I think it's okay. The contract of ConcurrentHashMap::computeIfAbsent is actually a bit stronger than ConcurrentMap::computeIfAbsent, so it guarantees that the mapping function will be invoked at most once per key.

However, looking at this again, doesn't this mean that two calls for the same part of a blob happening concurrently will be returned the same IndexInput instance? That's not safe, because IndexInputs are stateful and not thread safe:

> IndexInput may only be used from one thread, because it is not thread safe (it keeps internal state like file position).

Reply:
Yep that is a good catch! I thought IndexInput is cloned somewhere but it is not so will clone it before using it